Improve social preview section in new post page

Will-Howard opened the pull request:
With fallback image:
![Screenshot 2023-01-25 at 14 13 46](https://user-images.githubusercontent.com/77623106/214586804-611f23b0-3a98-4e2b-a91f-a50038d4b557.png)
With uploaded image:
![Screenshot 2023-01-25 at 14 15 15](https://user-images.githubusercontent.com/77623106/214586813-546c5ac8-90c5-4244-9ee5-20e8cbd4e1cc.png)
Falling back to first image in post. (Note: in the designs this is removable, it was much simpler to make it non-removeable so that's how I've done it. If people think it's important I can make it removeable):
![Screenshot 2023-01-25 at 14 15 48](https://user-images.githubusercontent.com/77623106/214586820-01f99d0f-0a6a-4c2a-bef3-ccbbf4e97fa2.png)
Empty body:
![Screenshot 2023-01-25 at 14 19 01](https://user-images.githubusercontent.com/77623106/214587565-b071fedb-745b-4e5e-8d38-a61bea136353.png)
Dark mode:
![Screenshot 2023-01-30 at 15 49 10](https://user-images.githubusercontent.com/77623106/215525405-912487d6-5476-4177-a57b-7850df55e13b.png)
LessWrong light and dark mode:
![Screenshot 2023-01-30 at 15 51 04](https://user-images.githubusercontent.com/77623106/215525790-8f64d85c-8b8b-42ca-a98c-bef678fa2fd5.png)
![Screenshot 2023-01-30 at 15 50 34](https://user-images.githubusercontent.com/77623106/215525822-08d56045-8084-406e-86a2-9ad31472653d.png)




┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203821691637607) by [Unito](https://www.unito.io)

Will-Howard commented on packages/lesswrong/components/form-components/SocialPreviewUpload.tsx:
```diff
@@ -0,0 +1,160 @@
+ * 3. The head tags are set in PostsPage (packages/lesswrong/components/posts/PostsPage/PostsPage.tsx)
+ *  3.1 plaintextDescription is truncated
+ *  3.2 socialPreviewImageUrl is just used directly
+ */
```
It's worth reading this comment

s-cheng replied:
Thanks, this is super helpful!

s-cheng commented on packages/lesswrong/components/form-components/SocialPreviewUpload.tsx:
```diff
@@ -0,0 +1,160 @@
+  return {description: previewDesc, fallbackImageUrl: img?.getAttribute("src") || null}
+}
+
+const SocialPreviewUpload = ({name, document, updateCurrentValues, clearField, label, croppingAspectRatio, classes}: {
```
@Will-Howard Just FYI this component breaks the layout on mobile, prob want to at least wrap the text onto a separate line

Will-Howard replied:
This is fixed now

s-cheng replied:
I still think the text should move onto its own line, since the preview gets really small, plus it still breaks the layout for small screens.

<img width="339" alt="Screen Shot 2023-01-31 at 11 02 46 AM" src="https://user-images.githubusercontent.com/9057804/215815024-add7d785-d804-4d58-ae7f-011126fb1b4f.png">

s-cheng commented on packages/lesswrong/components/form-components/SocialPreviewUpload.tsx:
```diff
@@ -0,0 +1,160 @@
+ * is _almost_ exactly the same, but skips some overly complicated or potentially slow steps, namely:
+ * - trimming whitespace
+ * - rendering LaTex
+ * - handling markdown and DraftJS (seeing as these are no longer supported for non admins)
```
FYI all users can still select markdown, via an account setting

Will-Howard replied:
Thanks, made it work with markdown now 👍

s-cheng replied:
Nice thanks! Is this comment now out-of-date?

s-cheng commented on packages/lesswrong/components/form-components/ImageUpload2.tsx:
```diff
@@ -0,0 +1,177 @@
+    paddingTop: 10,
+    paddingBottom: 10,
+  },
+  imageBackground: {
+    backgroundSize: 'cover',
+    backgroundPosition: 'top right',
+    width: 370,
+    height: 195,
+    display: 'flex',
+  },
```
Is `top right` the right thing to do here? I tested previewing a tall image on Twitter via [this post](https://forum-staging.effectivealtruism.org/posts/ow9NB2tprTfymgXn2/hello-2) and it seems to center the image in the preview? We prob won't match the behavior of all social media platforms, but Twitter might be the most important one to match.

Also I think the size should be responsive, so it doesn't break the layout on mobile

s-cheng commented on packages/lesswrong/components/form-components/SocialPreviewUpload.tsx:
```diff
@@ -0,0 +1,160 @@
+      <div className={classes.blurb}>
+        A preview image makes it more likely that people will see your post.
+        <br/><br/>
+        If you're unsure which image to use, consider trying Unsplash or an AI image generator.
```
I think it would be useful to link to Unsplash here. Also it looks like the line height needs to be increased.

s-cheng commented on packages/lesswrong/components/form-components/ImageUpload2.tsx:
```diff
@@ -0,0 +1,177 @@
+        <script src="https://upload-widget.cloudinary.com/global/all.js" type="text/javascript" />
+      </Helmet>
+      <div className={classes.imageBackground} style={{ backgroundImage: `url(${imageUrl})` }}>
+        <div className={classes.buttonRow}>
+          <Button onClick={uploadWidget} className={classNames("image-upload-button", classes.button)}>
+            {imageId ? `Change` : `Upload ${label}`}
+          </Button>
+          {imageId && <Button className={classes.button} title="Remove" onClick={removeImg}>
+            Remove
+          </Button>}
```
I think the button colors here aren't quite right in dark mode (comparing with the design [here](https://www.figma.com/file/DF9TLXQJuPrEJYEesS05J0/Social-preview?node-id=81%3A69&t=ojEM6DTRCraDhnen-0)):

<img width="651" alt="Screen Shot 2023-01-27 at 1 31 13 AM" src="https://user-images.githubusercontent.com/9057804/215023900-440a2ddd-db97-4946-8707-350790ab2714.png">

s-cheng replied:
Also the design calls for a few other smaller tweaks, like changing the section border color and adding a border radius - I assume you're intentionally excluding the yellow points for now?

s-cheng reviewed (commented):
I haven't finished reviewing everything yet, but thought it might be helpful to publish these comments in case you want to address them on Friday

s-cheng commented on packages/lesswrong/components/editor/EditorFormComponent.tsx:
```diff
@@ -72,6 +72,19 @@ export const EditorFormComponent = ({form, formType, formProps, document, name,
+    // Preserve other fields in "contents" which may have been sent from the server
+    updateCurrentValues({[fieldName]: {...(document[fieldName] || {}), ...(await editorRef.current.submitData(null))}})
+  }, autosaveInterval, {leading: true}),
+  [saveBackup, autosaveInterval])
```
I think this is reasonable - I didn't notice any slowdown when typing, plus when using collab editing, things don't get saved in local storage anyway. Though I was testing with a collab editing doc and the preview image doesn't get populated properly when I open the editor to edit the doc (though I think it does get populated when the text changes).

Also I noticed this issue with the top bar showing too many things, which seems to not replicate on `master` but I'm not totally convinced is related to this branch. I saw it when I first shared the doc, and then later I saw it when I opened the share link in incognito and typed some things as a second user, then closed that window.
<img width="678" alt="Screen Shot 2023-01-27 at 11 16 01 AM" src="https://user-images.githubusercontent.com/9057804/215137394-2b393ca6-7e39-466a-ab4a-842470bdd400.png">

Will-Howard replied:
Thanks for catching that, it turned out to be a fairly nasty bug that I've had to do a workaround for, see:
```
  /**
   * This is a workaround for a bug resulting from the PresenceList and TrackChangesData plugins.
   * When getDataWithDiscardedSuggestions is called in Editor.tsx (called via throttledSetContentsValue)
   * it results in a new presence list being added with "0 connected users". I have no idea why this happens,
   * and both the plugins are closed source so it will be very hard to work out. This workaround removes all
   * but the first presence list, which is the one that actually contains the correct data. Any others that
   * are created temporarily are also hidden by the "& .ck-presence-list:nth-child(n+2)" selector above.
   */
  const presenceListChildren = presenceListRef.current?.querySelectorAll(".ck-presence-list");
  useEffect(() => {
    // remove all but the first presence list
    if (!presenceListChildren) return;

    const [_, ...otherChildren] = Array.from(presenceListChildren);
    for (let child of otherChildren) {
      child.remove();
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [presenceListChildren?.length])
```

s-cheng replied:
Oof

s-cheng commented on packages/lesswrong/components/form-components/ImageUpload2.tsx:
```diff
@@ -0,0 +1,177 @@
+      <Helmet>
+        <script src="https://upload-widget.cloudinary.com/global/all.js" type="text/javascript" />
+      </Helmet>
+      <div className={classes.imageBackground} style={{ backgroundImage: `url(${imageUrl})` }}>
```
I think this should not try to apply a background image if `imageUrl` is not defined. Also maybe we should have some default background (such as a color or pattern) instead of no background? I'm not sure, I would ask Agnes for a suggestion.

<img width="654" alt="Screen Shot 2023-01-27 at 3 05 34 PM" src="https://user-images.githubusercontent.com/9057804/215187216-b4eff5b3-2cf0-4a6c-b77f-37c2c84d70d7.png">

Will-Howard replied:
I have just made it fall back to a plain background for now (@agnesstenlund do you think this is ok?)
![Screenshot 2023-01-30 at 15 56 44](https://user-images.githubusercontent.com/77623106/215527380-3c57fdef-86f8-47b2-9850-f47811eddde9.png)

agnesstenlund replied:
yup I think white works well here 👍 I guess the truth is it would be the logo if it's empty, but this is likely more clear

s-cheng commented on packages/lesswrong/components/form-components/ImageUpload2.tsx:
```diff
@@ -0,0 +1,177 @@
+    setImageId(null)
+  }
+
+  const [imageId, setImageId] = useState(() => {
+    if (document && document[name]) {
+      return document[name];
+    }
+    return ''
+  })
```
I noticed that this is not properly being set when editing an existing post. It looks like `document.socialPreviewImageId` is `undefined` even if I've manually added a social preview image previously, so maybe we need to update a fragment or something?

s-cheng commented on packages/lesswrong/components/form-components/SocialPreviewUpload.tsx:
```diff
@@ -0,0 +1,160 @@
+    };
+  }
+
+  const html = dataWithDiscardedSuggestions ?? originalContents.data
```
Should we be defaulting to `dataWithDiscardedSuggestions`? I'm kind of unsure but I guess that's the right thing to do here, to ignore suggestions when building the preview data. (Unrelated to this PR, but I think this is a really bad variable name, since I keep thinking that it's the opposite of what it's meant to be, but I guess it's just to [match CKEditor's naming](https://ckeditor.com/docs/ckeditor5/latest/api/module_track-changes_trackchangesdata-TrackChangesData.html#function-getDataWithDiscardedSuggestions))

Will-Howard replied:
This is what we use on the backend to create the preview description so I think it makes sense to do so here

s-cheng reviewed (changes requested):
This is awesome, so excited to have it! :D And excited to replace our old ImageUpload component - this is way better. There are a few functionality thing to address. Let me know when it's ready for review again, and I'll test it out some more.

Will-Howard commented on packages/lesswrong/lib/collections/posts/formGroups.ts:
```diff
@@ -15,17 +15,35 @@ export const formGroups: Partial<Record<string,FormGroup>> = {
     order: 21,
     label: "Event Details"
   },
+  socialPreview: {
+    name: "socialPreview",
+    order: 22,
+    label: "Edit Link Preview",
+    startCollapsed: false,
+  },
```
👍 I've disabled it for both questions and events, there is currently no way to set `startCollapsed` based on the props and I didn't think it was worth adding just for this

s-cheng replied:
That sounds good to me, though I still see it when creating a new event and a new question.

agnesstenlund commented:
this is really really great! 👏  I wasn't able to run it properly locally (will ping you about help with that) but the screenshots look great, thanks for the attention to detail

Only thing I noticed is I think the link preview should be below the Post topics instead of above.

> in the designs this is removable, it was much simpler to make it non-removeable so that's how I've done it. If people think it's important I can make it removeable
I agree making it removeable doesn't seem too important. What I would do though is change wording here to `Change preview image` instead of Upload (so they don't think it's not already "done")

agnesstenlund commented:
After checking out the branch I noticed two small things:

* Hovering the button should be styled inverse–so more transparent as standard state (when not hovering), and then darker when hovering 🙏 

* Not sure if mentioned above but these little dots that gets added in the body text of the preview looks a little funky
![Screenshot 2023-01-30 at 19 16 55](https://user-images.githubusercontent.com/13235378/215561693-d50b3c21-678c-4962-9927-f5bf433f4871.png)

Will-Howard commented:
@agnesstenlund Thanks for the review, I've changed these things:
 - Moved it below post topics
 - Changed wording to "Change Preview Image" in the case where it's using the one from the post
 - Inverted the normal/hover styling for the buttons

> Not sure if mentioned above but these little dots that gets added in the body text of the preview looks a little funky
Hmm I actually quite like the little dots, I feel there has to be some way of indicating line breaks without actually adding a line break. I'm open to other suggestions though

s-cheng commented on packages/lesswrong/components/form-components/SocialPreviewUpload.tsx:
```diff
@@ -0,0 +1,181 @@
+      <div className={classes.blurb}>
+        A preview image makes it more likely that people will see your post.
+        <br/><br/>
+        If you're unsure which image to use, consider trying <a href="https://unsplash.com/">Unsplash</a> or an AI image generator.
```
Can we open this link in a new tab, so as not to interrupt the editing process?

s-cheng reviewed (changes requested):
@Will-Howard Thanks so much for addressing my comments! I just left a few more, plus I see Agnes left one as well. I mostly want to make sure that we hide it properly for LW before approving this.

Will-Howard commented:
> would it be possible to simply show the line breaks?

Ah I was a bit confused about what was going on here because I hadn't changed anything to do with the dots in this PR. It turns out when I upgraded the `html-to-text` library it changed the defaults about how many new lines are added between headings and paragraphs in such a way that dots were added when they weren't before (for both this component and the actual preview that would appear in e.g. slack).

I've changed it to be back to how it was before now, so a single newline is added between headings and paragraphs, and dots are added between two paragraphs

s-cheng reviewed (approved):
🎉

PatMyron commented:
For link posts, it could be awesome to automatically default to the open graph image from the underlying link without everyone having to actively upload it
